pip/req/req_set.py

```diff
@@ -228,7 +232,11 @@ def prepare_files(self, finder):
             if not self.ignore_installed and not req_to_install.editable:
                 req_to_install.check_if_exists()
                 if req_to_install.satisfied_by:
-                    if self.upgrade:
+                    # check if we are upgrading and that we don't already have
+                    # an exact version match
+                    if self.upgrade and not (
+                            req_to_install.satisfied_by.as_requirement() ==
+                            req_to_install.req):
```

Comment:
In order to make the code nicer and to workaround what I think is a setuptools problem, I changed it to this:

```diff
--- a/pip/req/req_set.py
+++ b/pip/req/req_set.py
@@ -228,7 +228,12 @@ class RequirementSet(object):
             if not self.ignore_installed and not req_to_install.editable:
                 req_to_install.check_if_exists()
                 if req_to_install.satisfied_by:
-                    if self.upgrade:
+                    # check if we are upgrading and that we don't already have
+                    # an exact version match
+                    satisfied = (
+                        req_to_install.satisfied_by.as_requirement() ==
+                        req_to_install.req)
+                    if self.upgrade and not satisfied:
                         if not self.force_reinstall and not req_to_install.url:
                             try:
                                 url = finder.find_requirement(
```